File: src/main/java/uk/gov/ons/ssdc/caseprocessor/messaging/NewCaseReceiver.java
package uk.gov.ons.ssdc.caseprocessor.messaging;

import static uk.gov.ons.ssdc.caseprocessor.rasrm.constants.RasRmConstants.BUSINESS_SAMPLE_DEFINITION_URL_SUFFIX;
import static uk.gov.ons.ssdc.caseprocessor.utils.JsonHelper.convertJsonBytesToEvent;

import com.fasterxml.jackson.core.JsonProcessingException;
import com.fasterxml.jackson.databind.ObjectMapper;
import java.sql.SQLException;
import java.time.OffsetDateTime;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;
import java.util.Map;
import java.util.Set;
import java.util.UUID;
import java.util.stream.Collectors;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.integration.annotation.MessageEndpoint;
import org.springframework.integration.annotation.ServiceActivator;
import org.springframework.messaging.Message;
import org.springframework.transaction.annotation.Transactional;
import uk.gov.ons.ssdc.caseprocessor.logging.EventLogger;
import uk.gov.ons.ssdc.caseprocessor.model.dto.CaseScheduledTask;
import uk.gov.ons.ssdc.caseprocessor.model.dto.CaseScheduledTaskGroup;
import uk.gov.ons.ssdc.caseprocessor.model.dto.EventDTO;
import uk.gov.ons.ssdc.caseprocessor.model.dto.NewCase;
import uk.gov.ons.ssdc.caseprocessor.model.dto.ScheduleTemplate;
import uk.gov.ons.ssdc.caseprocessor.model.dto.ScheduleTemplateTask;
import uk.gov.ons.ssdc.caseprocessor.model.dto.ScheduleTemplateTaskGroup;
import uk.gov.ons.ssdc.caseprocessor.model.repository.CaseRepository;
import uk.gov.ons.ssdc.caseprocessor.model.repository.CollectionExerciseRepository;
import uk.gov.ons.ssdc.caseprocessor.rasrm.service.RasRmCaseNotificationEnrichmentService;
import uk.gov.ons.ssdc.caseprocessor.service.CaseService;
import uk.gov.ons.ssdc.caseprocessor.service.ScheduledTaskService;
import uk.gov.ons.ssdc.caseprocessor.utils.CaseRefGenerator;
import uk.gov.ons.ssdc.common.model.entity.Case;
import uk.gov.ons.ssdc.common.model.entity.CollectionExercise;
import uk.gov.ons.ssdc.common.model.entity.EventType;
import uk.gov.ons.ssdc.common.model.entity.Survey;
import uk.gov.ons.ssdc.common.validation.ColumnValidator;

@MessageEndpoint
public class NewCaseReceiver {
  private final CaseRepository caseRepository;
  private final CaseService caseService;
  private final CollectionExerciseRepository collectionExerciseRepository;
  private final EventLogger eventLogger;
  private final RasRmCaseNotificationEnrichmentService rasRmNewBusinessCaseEnricher;
  private final ScheduledTaskService scheduledTaskService;

  @Value("${caserefgeneratorkey}")
  private byte[] caserefgeneratorkey;

  public NewCaseReceiver(
      CaseRepository caseRepository,
      CaseService caseService,
      CollectionExerciseRepository collectionExerciseRepository,
      EventLogger eventLogger,
      RasRmCaseNotificationEnrichmentService rasRmNewBusinessCaseEnricher,
      ScheduledTaskService scheduledTaskService) {
    this.caseRepository = caseRepository;
    this.caseService = caseService;
    this.collectionExerciseRepository = collectionExerciseRepository;
    this.eventLogger = eventLogger;
    this.rasRmNewBusinessCaseEnricher = rasRmNewBusinessCaseEnricher;
    this.scheduledTaskService = scheduledTaskService;
  }

  @Transactional
  @ServiceActivator(inputChannel = "newCaseInputChannel", adviceChain = "retryAdvice")
  public void receiveNewCase(Message<byte[]> message) throws JsonProcessingException, SQLException {
    EventDTO event = convertJsonBytesToEvent(message.getPayload());

    NewCase newCasePayload = event.getPayload().getNewCase();

    if (caseRepository.existsById(newCasePayload.getCaseId())) {
      // Case already exists, so let's not overwrite it... swallow the message quietly
      return;
    }

    CollectionExercise collex =
        collectionExerciseRepository
            .findById(newCasePayload.getCollectionExerciseId())
            .orElseThrow(
                () ->
                    new RuntimeException(
                        "Collection exercise '"
                            + newCasePayload.getCollectionExerciseId()
                            + "' not found"));

    ColumnValidator[] columnValidators = collex.getSurvey().getSampleValidationRules();
    checkNewSampleWithinSampleDefinition(columnValidators, newCasePayload);
    checkNewSensitiveWithinSampleSensitiveDefinition(columnValidators, newCasePayload);

    validateNewCase(newCasePayload, columnValidators);

    Map<String, String> sample = newCasePayload.getSample();

    if (collex
        .getSurvey()
        .getSampleDefinitionUrl()
        .endsWith(BUSINESS_SAMPLE_DEFINITION_URL_SUFFIX)) {
      sample =
          rasRmNewBusinessCaseEnricher.notifyRasRmAndEnrichSample(sample, collex.getMetadata());
    }

    Case newCase = new Case();
    newCase.setId(newCasePayload.getCaseId());
    newCase.setCollectionExercise(collex);

    List<CaseScheduledTaskGroup> caseScheduledTaskGroups = getSchedule(collex.getSurvey());

    newCase.setSchedule(caseScheduledTaskGroups);

    newCase.setSample(sample);
    newCase.setSampleSensitive(newCasePayload.getSampleSensitive());

    newCase = saveNewCaseAndStampCaseRef(newCase);
    caseService.emitCaseUpdate(
        newCase, event.getHeader().getCorrelationId(), event.getHeader().getOriginatingUser());

    if (caseScheduledTaskGroups != null) {
      scheduledTaskService.createScheduledTasksFromSchedulePlan(
          caseScheduledTaskGroups, newCase.getId());
    }

    eventLogger.logCaseEvent(newCase, "New case created", EventType.NEW_CASE, event, message);
  }

  private Set<String> checkNewSensitiveWithinSampleSensitiveDefinition(
      ColumnValidator[] columnValidators, NewCase newCasePayload) {
    Set<String> sensitiveColumns =
        Arrays.stream(columnValidators)
            .filter(ColumnValidator::isSensitive)
            .map(ColumnValidator::getColumnName)
            .collect(Collectors.toSet());
    if (!sensitiveColumns.containsAll(newCasePayload.getSampleSensitive().keySet())) {
      throw new RuntimeException(
          "Attempt to send sensitive data to RM which was not part of defined sample");
    }

    return sensitiveColumns;
  }

  private void checkNewSampleWithinSampleDefinition(
      ColumnValidator[] columnValidators, NewCase newCasePayload) {
    Set<String> nonSensitiveColumns =
        Arrays.stream(columnValidators)
            .filter(columnValidator -> !columnValidator.isSensitive())
            .map(ColumnValidator::getColumnName)
            .collect(Collectors.toSet());
    if (!nonSensitiveColumns.containsAll(newCasePayload.getSample().keySet())) {
      throw new RuntimeException("Attempt to send data to RM which was not part of defined sample");
    }
  }

  private void validateNewCase(NewCase newCasePayload, ColumnValidator[] columnValidators) {
    List<String> validationErrors = new ArrayList<>();

    for (ColumnValidator columnValidator : columnValidators) {
      if (columnValidator.isSensitive()) {
        columnValidator
            .validateRow(newCasePayload.getSampleSensitive(), true)
            .ifPresent(validationErrors::add);
      } else {
        columnValidator
            .validateRow(newCasePayload.getSample(), true)
            .ifPresent(validationErrors::add);
      }
    }

    if (!validationErrors.isEmpty()) {
      throw new RuntimeException(
          "NEW_CASE event: "
              + validationErrors.stream().collect(Collectors.joining(System.lineSeparator())));
    }
  }

  private Case saveNewCaseAndStampCaseRef(Case caze) {
    Case newCase = caseRepository.saveAndFlush(caze);
    newCase.setCaseRef(
        CaseRefGenerator.getCaseRef(newCase.getSecretSequenceNumber(), caserefgeneratorkey));
    return caseRepository.saveAndFlush(newCase);
  }

  // Hove this off into a process,  and this function to orchastrate not chain
  private List<CaseScheduledTaskGroup> getSchedule(Survey survey) throws JsonProcessingException {
    ObjectMapper objectMapper = new ObjectMapper();

    // needs something like this, will do for now.
    if (survey.getScheduleTemplate() == null
        || survey.getScheduleTemplate().toString().length() == 0) {
      return null;
    }

    ScheduleTemplate scheduleTemplate =
        objectMapper.readValue((String) survey.getScheduleTemplate(), ScheduleTemplate.class);

    return createSchedule(scheduleTemplate);
  }

  private List<CaseScheduledTaskGroup> createSchedule(ScheduleTemplate scheduleTemplate) {

    List<CaseScheduledTaskGroup> caseScheduledTaskGroups = new ArrayList<>();
    // why add 5 seconds, well because of ATs.
    // This certainly isn't perfect
    // TODO: remove test crutch in production code.
    OffsetDateTime caseTaskGroupStartDate = OffsetDateTime.now().plusSeconds(5);

    for (ScheduleTemplateTaskGroup scheduleTemplateTaskGroup :
        scheduleTemplate.getScheduleTemplateTaskGroups()) {
      CaseScheduledTaskGroup caseScheduledTaskGroup = new CaseScheduledTaskGroup();
      caseScheduledTaskGroup.setName(scheduleTemplateTaskGroup.getName());

      caseScheduledTaskGroup.setDateOffSet(scheduleTemplateTaskGroup.getDateOffSet());

      caseScheduledTaskGroups.add(caseScheduledTaskGroup);

      caseTaskGroupStartDate =
          caseTaskGroupStartDate.plusSeconds(
              caseScheduledTaskGroup.getDateOffSet().getDateUnit().getDuration().getSeconds()
                  * caseScheduledTaskGroup.getDateOffSet().getOffset());

      caseScheduledTaskGroup.setScheduledTasks(
          createScheduledTaskList(scheduleTemplateTaskGroup, caseTaskGroupStartDate));
    }

    return caseScheduledTaskGroups;
  }

  private List<CaseScheduledTask> createScheduledTaskList(
      ScheduleTemplateTaskGroup scheduleTemplateTaskGroup, OffsetDateTime startDate) {
    List<CaseScheduledTask> caseScheduledTaskList = new ArrayList<>();

    for (ScheduleTemplateTask scheduleTemplateTask :
        scheduleTemplateTaskGroup.getScheduleTemplateTasks()) {
      CaseScheduledTask caseScheduledTask = new CaseScheduledTask();
      caseScheduledTask.setId(UUID.randomUUID());
      caseScheduledTask.setName(scheduleTemplateTask.getName());
      caseScheduledTask.setScheduledTaskType(scheduleTemplateTask.getScheduledTaskType());
      caseScheduledTask.setPackCode(scheduleTemplateTask.getPackCode());

      OffsetDateTime newTime =
          startDate.plusSeconds(
              scheduleTemplateTask.getDateOffSet().getDateUnit().getDuration().getSeconds()
                  * scheduleTemplateTask.getDateOffSet().getOffset());

      caseScheduledTask.setRmScheduledDateTime(newTime);
      caseScheduledTask.setScheduledDateAsString(newTime.toString());

      caseScheduledTask.setDateOffSet(scheduleTemplateTask.getDateOffSet());
      caseScheduledTask.setEventIds(new ArrayList<>());
      caseScheduledTask.setUacsIds(new ArrayList<>());

      caseScheduledTaskList.add(caseScheduledTask);
    }

    return caseScheduledTaskList;
  }
}

File: python spike/GCP PubSub/caseprocessor/new_case_receiver.py
import json
from dataclass_wizard import fromdict
from caseprocessor.config import PubsubConfig
from caseprocessor.dto.event_dto import EventDTO
from caseprocessor.db.case_repository import CasesTable
from caseprocessor.db.collection_exercise_repository import CollectionExerciseTable
from caseprocessor.db.db_utility import create_session
from caseprocessor.validation.column_validator import ColumnValidator
from caseprocessor.entity.case import Case
from caseprocessor.case_ref_generator import get_case_ref

class NewCaseReceiver:
    __case_ref_generator_key = bytes("abc123", "utf-8")

    @classmethod
    def receive_new_case(cls, message: bytes):
        try:
            parsed_json = json.loads(message)
            event = fromdict(EventDTO, parsed_json)

            new_case_payload = event.payload.new_case

            print("----Json Deserialized----")
            print(new_case_payload)

            session = create_session()

            if CasesTable.exists_by_id(session, new_case_payload.case_id):
                return

            collex = CollectionExerciseTable.find_by_id(session, new_case_payload.collection_exercise_id)

            if collex is None:
                raise Exception("Collection exercise '"
                                + new_case_payload.collection_exercise_id()
                                + "' not found")

            column_validator = collex.survey.sample_validation_rules

            cls.__check_new_sensitive_within_sample_sensitive_definition(column_validator, new_case_payload)
            cls.__check_new_sample_within_sample_definition(column_validator, new_case_payload)
            cls.__validate_new_case(column_validator, new_case_payload)

            sample = new_case_payload.sample

            new_case = Case(
                case_id=new_case_payload.case_id,
                collection_exercise=collex,
                sample=new_case_payload.sample,
                sample_sensitive=new_case_payload.sample_sensitive
            )

            new_case = cls.__save_new_case_and_stamp_case_ref(new_case)

            # TODO: add logger and finish this off



        except TypeError as e:
            # TODO: throw error if json is wrong/can't cast to class
            print(f"something went wrong:\n{e}")

    @classmethod
    def __check_new_sensitive_within_sample_sensitive_definition(cls, column_validators, new_case_payload):
        sensitive_columns = {ColumnValidator.get_column_name(column_validator)
                             for column_validator in column_validators
                             if ColumnValidator.is_sensitive(column_validator)}

        if sensitive_columns.issubset(new_case_payload.sample_sensitive.keys()):
            raise Exception("Attempt to send sensitive data to RM which was not part of defined sample")

        return sensitive_columns

    @classmethod
    def __check_new_sample_within_sample_definition(cls, column_validators, new_case_payload):
        sensitive_columns = {ColumnValidator.get_column_name(column_validator)
                             for column_validator in column_validators
                             if not ColumnValidator.is_sensitive(column_validator)}

        if sensitive_columns.issubset(new_case_payload.sample_sensitive.keys()):
            raise Exception("Attempt to send data to RM which was not part of defined sample")

        return sensitive_columns

    @classmethod
    def __validate_new_case(cls, column_validators, new_case_payload):
        validate_errors = []

        for column_validator in column_validators:
            if column_validator.is_sensitive:
                error = column_validator.validate_row(new_case_payload.sample_sensitive,
                                                      exclude_data_from_returned_error_msgs=True)
                if error is not None:
                    validate_errors.append(error)
            else:
                error = column_validator.validate_row(new_case_payload.sample,
                                                      exclude_data_from_returned_error_msgs=True)
                if error is not None:
                    validate_errors.append(error)

        if validate_errors:
            raise Exception("NEW_CASE event: " + "\n.join(validate_errors)")

    @classmethod
    def __save_new_case_and_stamp_case_ref(cls, caze, session):
        CasesTable.save_and_flush(session, caze)
        caze.case_ref = get_case_ref(caze.secret_sequence_number(), cls.__case_ref_generator_key)
        return caze
